[10.0][ADD] account_payment_mode_auto_reconcile

grindtildeath opened the pull request:
This module adds a checkbox 'auto_reconcile_outstanding_credits' on account
payment modes to allow automatic reconciliation on account invoices if it is
checked.

Automatic reconciliation of outstanding credits will only happen on customer
invoices at validation if the payment mode is set or when the payment mode is
changed on an open invoice.

grindtildeath committed:
Add account_payment_mode_auto_reconcile

grindtildeath committed:
Override write instead of using onchange

grindtildeath committed:
Add missing utf-8 comment

grindtildeath committed:
Fix warning message

grindtildeath committed:
Restrict to customer invoices

grindtildeath committed:
Add readme and load tests

JordiBForgeFlow commented:
Could you tell me what is the use case? Why would one auto-reconcile at the time of validating the invoice, with no control on what to reconcile against? Thanks!

grindtildeath commented:
@jbeficent The company is using Direct Debit orders to get payments from selected customers and the idea is to reconcile any existing credit notes for such customers before the invoice is sent and the debit order is created. We also backported v12.0 report to print the invoices with payments (https://github.com/OCA/account-invoicing/pull/567) so that their customers receive the amount that will be debited.

JordiBForgeFlow commented:
Thanls for the clarification. Will review the PR

i-vyshnevska commented on account_payment_mode_auto_reconcile/models/account_invoice.py:
```diff
@@ -0,0 +1,130 @@
+            )
+            # Get outstanding credits in chronological order
+            # (using reverse because aml is sorted by date desc as default)
+            credits = credits_info.get('content')
```
seems credits are reserved word

i-vyshnevska commented on account_payment_mode_auto_reconcile/models/account_invoice.py:
```diff
@@ -0,0 +1,130 @@
+
+    @api.multi
+    def write(self, vals):
+        res = super(AccountInvoice, self).write(vals)
```
shouldn't  it would be performed before write?

grindtildeath replied:
We can't modify `vals` to do the reconciliation, so better do it afterwards

i-vyshnevska commented on account_payment_mode_auto_reconcile/models/account_invoice.py:
```diff
@@ -0,0 +1,130 @@
+                    invoice.auto_reconcile_credits(
+                        partial_allowed=partial_allowed
+                    )
+                elif invoice.payment_move_line_ids:
```
some comments pls, seems we call auto_unreconcile_credits for  default payment_mode is it correct?

i-vyshnevska replied:
"is not using auto reconcile" we don't check it here, we drop lines always

grindtildeath replied:
If we get there, it means the payment mode is not set to auto-reconcile (otherwise it would pass the `if` above)

i-vyshnevska replied:
exactly, mode is not in 'auto-reconcile', so how do we now was this lines added by auto reconcile or manually?
this will applly one policy for all invoices

grindtildeath replied:
I thought I'd leave it so, but now fixed in f519930

i-vyshnevska commented on account_payment_mode_auto_reconcile/tests/test_partner_auto_reconcile.py:
```diff
@@ -0,0 +1,151 @@
+        })
+        self.assertTrue(self.payment_mode.auto_reconcile_outstanding_credits)
+        self.assertEqual(self.invoice_copy.residual, 1500)
+        auto_rec_invoice.action_invoice_open()
```
should this behavior run when we go from open -> done?

grindtildeath replied:
I don't really understand your question. The auto reconciliation will be done when the invoice is open, and if there are enough outstanding credit to reconcile, the invoice will be set to done...

i-vyshnevska replied:
if invoice wasn't fully reconciled and after new lines appeared, they will not be added here
if this action should performed only once than ok,
 maybe it can be done in action_invoice_validate and unreconcilation in action_invoice_cancel instead of write method

grindtildeath replied:
As written in the description "Automatic reconciliation of outstanding credits will only happen on customer
invoices at validation if the payment mode is set or when the payment mode is
changed on an open invoice."

jcoux commented on account_payment_mode_auto_reconcile/models/account_invoice.py:
```diff
@@ -0,0 +1,130 @@
+                    'Validating invoices with this payment mode will reconcile'
+                    ' any outstanding credits.'
+                )
+                invoice.display_payment_mode_warning = True
```
Already done line 98.

jcoux commented on account_payment_mode_auto_reconcile/models/account_invoice.py:
```diff
@@ -0,0 +1,130 @@
+                invoice.payment_mode_warning = _(
+                    'Changing payment mode will unreconcile existing payments.'
+                )
+                invoice.display_payment_mode_warning = True
```
Already done line 98.

jcoux commented on account_payment_mode_auto_reconcile/models/account_invoice.py:
```diff
@@ -0,0 +1,130 @@
+                invoice.payment_mode_warning = _(
+                    'Changing payment mode will reconcile outstanding credits.'
+                )
+                invoice.display_payment_mode_warning = True
```
Already done line 98.

jcoux reviewed (commented):
Just technical review.
I don't know if functionally it's good or not.

TDu commented on account_payment_mode_auto_reconcile/models/account_payment_mode.py:
```diff
@@ -0,0 +1,19 @@
+             "using this payment mode is validated, or when this payment mode "
+             "is defined on an open invoice."
+    )
+    auto_reconcile_allow_partial = fields.Boolean(
```
There is no mention of this in the description. It could help

grindtildeath replied:
Attended in 1e2ebca

grindtildeath committed:
Improvements after reviews

jcoux reviewed (approved):
Technically good for me :+1:

grindtildeath committed:
Improve doc

grindtildeath committed:
Unreconcile only automatically reconciled payments

i-vyshnevska commented on account_payment_mode_auto_reconcile/tests/test_partner_auto_reconcile.py:
```diff
@@ -0,0 +1,193 @@
+
+    @classmethod
+    def setUpClass(cls):
+        super(TestPartnerAutoReconcile, cls).setUpClass()
```
cls.env = cls.env(context=dict(cls.env.context, tracking_disable=True))

grindtildeath committed:
fixup! Improvements after reviews

jbaudoux commented:
The communication in the direct debit contains the move line reference if not empty, otherwise it contains the move name. For invoices with direct debit, you would configure the communication as free. So the move line does not get a communication and the payment line will contain the invoice number (i.e. move name) which is the most meaningful for the customer as there are no debit advice sent to the bank.
When you deduct a credit note at invoice confirmation, you better notify the customer. Easiest would be to write in the invoice and move line reference the move name (invoice number) and list of credit notes that have been registered as payment. This will allow the customer to understand why the total amount of the invoice has not been debited.

vrenaville committed:
[FIX] typo in variable name

grindtildeath committed:
Merge pull request #1 from vrenaville/10.0_add_account_payment_mode_auto_reconcile

[FIX] typo in variable name

TDu committed:
Add auto reconcile only same journal (#2)

* Add auto reconcile only same journal

Add an option to auto reconcile only credits that are on the same
journal than the invoice.
This make sure that unrelated payment will not be used automatically.

* Update account_payment_mode_auto_reconcile/models/account_invoice.py

Co-Authored-By: Akim Juillerat <akim.juillerat@camptocamp.com>

* Update account_payment_mode_auto_reconcile/tests/test_partner_auto_reconcile.py

Co-Authored-By: Akim Juillerat <akim.juillerat@camptocamp.com>

TDu committed:
fixup! Add auto reconcile only same journal (#2)

sebalix commented on account_payment_mode_auto_reconcile/models/account_invoice.py:
```diff
@@ -82,6 +84,16 @@ def auto_reconcile_credits(self, partial_allowed=True):
+        line_ids = [credit['id'] for credit in credits_dict]
+        lines = self.env['account.move.line'].browse(line_ids).filtered(
+            lambda line: line.journal_id == self.journal_id
+        )
```
@TDu To get better performance (don't know how many records we could process here), could you rewrite this using `search` instead of `browse`+`filtered`? That way no prefetch will be done on the database as you are also only using `line.ids` right after.

TDu replied:
Thanks for your review, no compromise on performance, fixed.

TDu committed:
fixup! fixup! Add auto reconcile only same journal (#2)

TDu committed:
fixup! fixup! fixup! Add auto reconcile only same journal (#2)

i-vyshnevska commented on account_payment_mode_auto_reconcile/tests/test_partner_auto_reconcile.py:
```diff
@@ -0,0 +1,235 @@
+
+    def test_invoice_auto_reconcile_different_journal(self):
+        """Check not reconciling credits on different journal."""
+        self.payment_mode.auto_reconcile_same_journal = True
```
```suggestion
        self.payment_mode.auto_reconcile_same_journal = False
```
?

TDu replied:
No this is correct the test is checking that when the option is set credits on a different journal is not reconciled.

i-vyshnevska commented on account_payment_mode_auto_reconcile/models/account_payment_mode.py:
```diff
@@ -0,0 +1,26 @@
+        string="Allow partial",
+        help="Allows automatic partial reconciliation of outstanding credits",
+    )
+    auto_reconcile_same_journal = fields.Boolean(
```
instead of forcing reconciliation only under one journal it should fall to best match based on amount and date 
how for instance it will run reconciliation if under one journal there will be 2 payments for 200 and 300 with invoice on 300?
and forcing to one journal will exclude reconciliation of payment with different currencies, which can be useful, it's better to match partial payment in different currency than didn't match anything

TDu replied:
The truth is : this is a request from the consultants that have been thinking hard to please their customer. I implemented the solution they thought was the best fit.
Also this is only an option that can be activated or not !

pedrobaeza commented:
Shouldn't this go to account-reconcile repo?

grindtildeath committed:
FIX: Override invoice_validate instead of action_invoice_open

This small change is needed since action_invoice_open filters
invoices that can be open from other invoices that cannot.

Therefore using invoice_validate here allows for better
customizations by ensuring that the invoices to be handled are
really open (i.e. have a move generated).

grindtildeath commented:
@pedrobaeza I agree, I'll open a new PR there asap :wink:
